components/airtable/new-modified-or-deleted-records.js

```diff
@@ -0,0 +1,111 @@
+const airtable = require('https://github.com/PipedreamHQ/pipedream/components/airtable/airtable.app.js');
+const moment = require('moment');
+const axios = require('axios');
+const Bottleneck = require('bottleneck');
+
+const limiter = new Bottleneck({
+  minTime: 200, // 5 requets per second
+});
+const axiosRateLimiter = limiter.wrap(axios);
+
+module.exports = {
+  name: 'New, Modified or Deleted Records',
+  version: '0.0.2',
+  props: {
+    db: '$.service.db',
+    airtable,
+    baseId: { type: '$.airtable.baseId', appProp: 'airtable' },
+    tableId: { type: '$.airtable.tableId', baseIdProp: 'baseId' },
+    timer: {
+      type: '$.interface.timer',
+      default: {
+        intervalSeconds: 60 * 5,
+      },
+    },
+  },
+  async run(event) {
+    const { baseId, tableId, viewId } = this;
+    const metadata = {
+      baseId,
+      tableId,
+      viewId,
+    };
+
+    const config = {
+      url: `https://api.airtable.com/v0/${encodeURIComponent(this.baseId)}/${encodeURIComponent(this.tableId)}`,
+      params: {},
+      headers: {
+        Authorization: `Bearer ${this.airtable.$auth.api_key}`,
+      },
+    };
+
+    const lastTimestamp = this.db.get('lastTimestamp');
+    const prevAllRecordIds = this.db.get('prevAllRecordIds');
+
+    if (lastTimestamp) {
+      config.params.filterByFormula = `LAST_MODIFIED_TIME() > "${lastTimestamp}"`;
+    }
+    const timestamp = new Date().toISOString();
+    const { data } = await axios(config);
+
+    let allRecordIds = [],
+      newRecordsCount = 0,
+      modifiedRecordsCount = 0,
+      deletedRecordsCount = 0;
+
+    if (!data.records.length) {
+      delete config.params.filterByFormula;
+
+      while (allRecordIds.length === 0 || config.params.offset) {
+        const { data } = await axiosRateLimiter(config);
+        if (!data.records.length || data.records.length === 0) return;
+        allRecordIds = [...allRecordIds, ...data.records.map(record => record.id)];
+        if (data.offset) {
+          config.params.offset = data.offset;
+        } else {
+          delete config.params.offset;
+        }
+      }
+
+      if (prevAllRecordIds && prevAllRecordIds.length > allRecordIds.length) {
+        let deletedRecordIds = prevAllRecordIds.filter(prevRecord => !allRecordIds.includes(prevRecord));
+        for (let recordID of deletedRecordIds) {
+          deletedRecordsCount++;
+          deletedRecordObj = {
+            metadata,
+            type: 'record_deleted',
+            id: recordID,
+          };
+          this.$emit(deletedRecordObj, {
+            summary: 'record_deleted',
+            id: recordID,
+          });
+        }
+      }
+    } else {
```

Comment:
Even in this case, you could have records that were deleted correct? If I add one record and delete another record in the same 5-minute period, won't I hit this `else` block and only process the added record?

Reply:
Yes that's true. For my use case I just wanted to have either one event trigger so i can rebuild my Gatsby site on data change. Let me think about how to address this properly. Do you have any suggestions?

Reply:
Yeah in this case, I'd recommend _always_ running the logic to check for deleted records, instead of separating the logic that runs when `!data.records.length` or the `else` case.

Reply:
Yes deleted the if condition. It triggers on both events now.